mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
+        return failure();
+      SmallVector<ast::Expr *> args{*lhs, *rhs};
+
+      // Check if it is in rewrite section but not in the let statement
+      bool inRewriteSection = parserContext == ParserContext::Rewrite;
+      if (inRewriteSection &&
+          nativeOperatorContext == NativeOperatorContext::Generic)
+        return emitError(
+            "nodiscard rule for mul operator is applied in rewrite section");
+
+      lhs = inRewriteSection ? createBuiltinCall(curToken.getLoc(),
+                                                 builtins.mulRewrite, args)
+                             : createBuiltinCall(curToken.getLoc(),
+                                                 builtins.mulConstraint, args);
+      continue;
+    }
+    case Token::div: {
+      consumeToken();
+      auto rhs = parseExp2Log2Expr();
+      if (failed(rhs))
+        return failure();
+      SmallVector<ast::Expr *> args{*lhs, *rhs};
+
+      // Check if it is in rewrite section but not in the let statement
+      bool inRewriteSection = parserContext == ParserContext::Rewrite;
+      if (inRewriteSection &&
+          nativeOperatorContext == NativeOperatorContext::Generic)
+        return emitError(
+            "nodiscard rule for div operator is applied in rewrite section");
+
+      lhs = inRewriteSection ? createBuiltinCall(curToken.getLoc(),
+                                                 builtins.divRewrite, args)
+                             : createBuiltinCall(curToken.getLoc(),
+                                                 builtins.divConstraint, args);
+      continue;
+    }
+    case Token::mod: {
+      consumeToken();
+      auto rhs = parseExp2Log2Expr();
+      if (failed(rhs))
+        return failure();
+      SmallVector<ast::Expr *> args{*lhs, *rhs};
+      bool inRewriteSection = parserContext == ParserContext::Rewrite;
+      if (inRewriteSection &&
+          nativeOperatorContext == NativeOperatorContext::Generic)
+        return emitError(
+            "nodiscard rule for mod operator is applied in rewrite section");
+
+      lhs = inRewriteSection ? createBuiltinCall(curToken.getLoc(),
+                                                 builtins.modRewrite, args)
+                             : createBuiltinCall(curToken.getLoc(),
+                                                 builtins.modConstraint, args);
+      continue;
+    }
+    default:
+      return lhs;
+    }
+  }
+}
+
+FailureOr<ast::Expr *> Parser::parseExp2Log2Expr() {
+  FailureOr<ast::Expr *> expr = nullptr;
+
   switch (curToken.getKind()) {
-  case Token::add: {
+  case Token::log2: {
+    consumeToken();
+    consumeToken(Token::l_paren);
+    expr = parseAddSubExpr();
+    if (failed(expr))
+      return failure();
+
+    // Check if it is in rewrite section but not in the let statement
+    bool inRewriteSection = parserContext == ParserContext::Rewrite;
+    if (inRewriteSection &&
+        nativeOperatorContext == NativeOperatorContext::Generic)
```

Comment:
Can't we check that this isn't a `Let`?
```suggestion
        nativeOperatorContext != NativeOperatorContext::Let)
```